engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,990 @@
+package websockets
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	dpmock "github.com/onflow/flow-go/engine/access/rest/websockets/data_providers/mock"
+	connmock "github.com/onflow/flow-go/engine/access/rest/websockets/mock"
+	"github.com/onflow/flow-go/engine/access/rest/websockets/models"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type WsControllerSuite struct {
+	suite.Suite
+
+	logger   zerolog.Logger
+	wsConfig Config
+}
+
+func (s *WsControllerSuite) SetupTest() {
+	s.logger = unittest.Logger()
+	s.wsConfig = NewDefaultWebsocketConfig()
+}
+
+func TestWsControllerSuite(t *testing.T) {
+	suite.Run(t, new(WsControllerSuite))
+}
+
+// TestSubscribeRequest tests the subscribe to topic flow.
+// We emulate a request message from a client, and a response message from a controller.
+func (s *WsControllerSuite) TestSubscribeRequest() {
+	// It still fails when run with -race even though we don't share any state
+	// (I tried changing logger & config to be unique in each test)
+	//s.T().Parallel()
+
+	s.T().Run("Happy path", func(t *testing.T) {
+		t.Parallel()
+
+		conn, dataProviderFactory, dataProvider := newControllerMocks(t)
+		controller := NewWebSocketController(s.logger, s.wsConfig, conn, dataProviderFactory)
+
+		dataProviderFactory.
+			On("NewDataProvider", mock.Anything, mock.Anything, mock.Anything, mock.Anything).
+			Return(dataProvider, nil).
+			Once()
+
+		id := uuid.New()
+		done := make(chan struct{}, 1)
```

Comment:
since this is never written to, no need to add a buffer. that my be confusing for someone in the future who would assume it will be written to somewhere